diff --git a/package/system/fstools/Makefile b/package/system/fstools/Makefile
index 4a11081..556d4b0 100644
--- a/package/system/fstools/Makefile
+++ b/package/system/fstools/Makefile
@@ -8,14 +8,14 @@
 include $(TOPDIR)/rules.mk

 PKG_NAME:=fstools
-PKG_VERSION:=2016-04-30
+PKG_VERSION:=2016-05-12

 PKG_RELEASE=$(PKG_SOURCE_VERSION)

 PKG_SOURCE_PROTO:=git
 PKG_SOURCE_URL=$(OPENWRT_GIT)/project/fstools.git
 PKG_SOURCE_SUBDIR:=$(PKG_NAME)-$(PKG_VERSION)
-PKG_SOURCE_VERSION:=bb92c29fc4cd90d396fc79078eba75fcb01c92ae
+PKG_SOURCE_VERSION:=ed224f0b1f9cfb3f8c5766cba3a60343c0eda586
 PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION)-$(PKG_SOURCE_VERSION).tar.gz
 CMAKE_INSTALL:=1
