diff --git toolchain/uClibc/config-0.9.33.2/common toolchain/uClibc/config-0.9.33.2/common
index 457d21f..679d924 100644
--- toolchain/uClibc/config-0.9.33.2/common
+++ toolchain/uClibc/config-0.9.33.2/common
@@ -186,7 +186,7 @@ UCLIBC_HAS_TZ_FILE_READ_MANY=y
 # UCLIBC_HAS_UTMPX is not set
 UCLIBC_HAS_WCHAR=y
 UCLIBC_HAS_WORDEXP=y
-# UCLIBC_HAS_XATTR is not set
+UCLIBC_HAS_XATTR=y
 # UCLIBC_HAS_XLOCALE is not set
 UCLIBC_HAS___PROGNAME=y
 # UCLIBC_LINUX_MODULE_24 is not set
